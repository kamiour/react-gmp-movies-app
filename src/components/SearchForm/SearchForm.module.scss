@import 'src/scss/variables';

.searchformWrapper {
  width: 100%;
  max-width: 790px;
  margin-top: -35px;

  .searchformTitle {
    text-align: left;
    font-weight: 300;
    font-size: 40px;
    line-height: 49px;
    letter-spacing: 1px;
    text-transform: uppercase;
    color: $color_white;
    margin-bottom: 38px;
  }
  
  .searchform {
    display: flex;
    justify-content: center;
    align-items: center;

    > div {
      width: 100%;
      margin-right: 14px;
    }
    
    .searchformBtn {
      min-width: 233px;
    }
  }
}
