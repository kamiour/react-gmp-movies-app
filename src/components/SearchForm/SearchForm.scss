@import 'src/scss/variables';

.searchform-wrapper {
  width: 100%;
  max-width: 790px;
  margin-top: -35px;

  .searchform-title {
    text-align: left;
    font-weight: 300;
    font-size: 40px;
    line-height: 49px;
    letter-spacing: 1px;
    text-transform: uppercase;
    color: $color_white;
    margin-bottom: 38px;
  }
  
  .searchform {
    display: flex;
    justify-content: center;
    align-items: center;

    .form-field {
      width: 100%;
      margin-right: 14px;
    }
    
    .searchform-btn {
      min-width: 233px;
    }
  }
}
