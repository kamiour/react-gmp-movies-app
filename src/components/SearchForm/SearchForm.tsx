import { useId } from 'react';
import { useRouter } from 'next/router';
import { Formik, Form } from 'formik';
import TextField from '../TextField/TextField';
// import './SearchForm.scss';

export default function SearchForm() {
  const router = useRouter();
  const { search: searchQuery } = router.query;
  const inputIdPrefix = useId();

  const handleSubmit = ({ searchValue }: { searchValue: string }) => {
    if (searchValue) {
      router.pathname = '/search/[search]';
      router.query.search = searchValue.toLocaleLowerCase();
    } else {
      router.pathname = '/search';
      delete router.query.search;
    }

    router.push(
      {
        pathname: router.pathname,
        query: router.query,
      },
      undefined,
      { shallow: true }
    );
  };

  return (
    <div className="searchform-wrapper">
      <h1 className="searchform-title">Find your movie</h1>

      <Formik initialValues={{ searchValue: (searchQuery as string) || '' }} onSubmit={handleSubmit}>
        <Form className="searchform">
          <TextField
            name="searchValue"
            className="form-input"
            id={`${inputIdPrefix}_search-input`}
            label=""
            type="text"
            placeholder="What do you want to watch?"
          />

          <button className="app-btn searchform-btn" type="submit">
            Search
          </button>
        </Form>
      </Formik>
    </div>
  );
}
