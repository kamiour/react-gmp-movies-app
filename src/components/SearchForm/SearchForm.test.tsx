import { render, waitFor } from '@testing-library/react';
import { BrowserRouter as Router, MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import SearchForm from './SearchForm';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUseNavigate,
}));

describe('SearchForm', () => {
  it('should match SearchForm snapshot', () => {
    const { asFragment } = render(
      <Router>
        <SearchForm />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should navigate to a search result on form submit', async () => {
    const { getByPlaceholderText, getByRole } = render(
      <MemoryRouter initialEntries={['/search?sortBy=release_date']}>
        <SearchForm />
      </MemoryRouter>
    );

    const searchInput = getByPlaceholderText('What do you want to watch?');
    expect(searchInput).toBeInTheDocument();
    userEvent.type(searchInput, 'Moon');

    const searchButton = getByRole('button');
    expect(searchButton).toBeInTheDocument();
    userEvent.click(searchButton);

    await waitFor(() => {
      expect(mockedUseNavigate).toBeCalledWith(
        {
          pathname: `/search/moon`,
          search: 'sortBy=release_date',
        },
        { replace: true }
      );
    });
  });
});
