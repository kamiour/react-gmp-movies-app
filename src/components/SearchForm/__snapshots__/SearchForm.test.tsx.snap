// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match SearchForm snapshot 1`] = `
<DocumentFragment>
  <div
    class="searchform-wrapper"
  >
    <h1
      class="searchform-title"
    >
      Find your movie
    </h1>
    <form
      action="#"
      class="searchform"
    >
      <div
        class="form-field"
      >
        <input
          class="form-input"
          id=":r0:_search-input"
          name="searchValue"
          placeholder="What do you want to watch?"
          type="text"
          value=""
        />
      </div>
      <button
        class="app-btn searchform-btn"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
</DocumentFragment>
`;
